src/qt/renderer.cpp

```diff
@@ -0,0 +1,634 @@
+                              wxDC& dc,
+                              const wxRect& rect,
+                              int flags = 0) override;
+
+    virtual wxSize GetCheckMarkSize(wxWindow *win) override;
+
+    virtual void DrawPushButton(wxWindow *win,
+                                wxDC& dc,
+                                const wxRect& rect,
+                                int flags = 0) override;
+
+    virtual void DrawItemSelectionRect(wxWindow *win,
+                                       wxDC& dc,
+                                       const wxRect& rect,
+                                       int flags = 0) override;
+
+    virtual void DrawChoice(wxWindow* win,
+                            wxDC& dc,
+                            const wxRect& rect,
+                            int flags = 0) override;
+
+    virtual void DrawComboBox(wxWindow* win,
+                              wxDC& dc,
+                              const wxRect& rect,
+                              int flags = 0) override;
+
+    virtual void DrawGauge(wxWindow* win,
+                           wxDC& dc,
+                           const wxRect& rect,
+                           int value,
+                           int max,
+                           int flags = 0);
+
+    virtual void DrawRadioBitmap(wxWindow* win,
+                                wxDC& dc,
+                                const wxRect& rect,
+                                int flags = 0 ) override;
+
+    virtual void DrawFocusRect(wxWindow* win,
+                               wxDC& dc,
+                               const wxRect& rect,
+                               int flags = 0) override;
+
+    virtual wxSize GetCheckBoxSize(wxWindow *win, int flags = 0) override;
+
+    virtual wxSplitterRenderParams GetSplitterParams(const wxWindow *win) override;
+
+    // Common code for DrawComboBox() and DrawChoice()
+    void DoDrawComboBox(wxWindow* win,
+                        wxDC& dc,
+                        const wxRect& rect,
+                        int flags, bool editable = true);
+};
+
+// ============================================================================
+// implementation
+// ============================================================================
+
+/* static */
+wxRendererNative& wxRendererNative::GetDefault()
+{
+    static wxRendererQt s_rendererQt;
+
+    return s_rendererQt;
+}
+
+static QPainter* wxGetQtPainter(const wxDC& dc)
+{
+    return static_cast<QPainter*>(dc.GetHandle());
+}
+
+// ----------------------------------------------------------------------------
+// list/tree controls drawing
+// ----------------------------------------------------------------------------
+
+int
+wxRendererQt::DrawHeaderButton(wxWindow *win,  wxDC& dc, const wxRect& rect, int flags,
+                               wxHeaderSortIconType sortArrow, wxHeaderButtonParams* params)
+{
+    auto painter = wxGetQtPainter(dc);
+    if ( !painter )
```

Comment:
Is it expected for this to happen or should it be a `wxCHECK()` (here and below)?

Reply:
AFAICT, no, the painter is not expected to be **NULL**. I'll update these and use  `wxCHECK()` instead...

Reply:
I just want to confirm that the painter can only be null if the DC is a `wxMemoryDC` with no bitmap selected. So the added checks are safe for the other types of `wxDC`s and should warn for the `wxMemoryDC` case mentioned above.